### PR TITLE
CDAP-4454: Schema validation for sources

### PR DESCRIPTION
**Pull request by jwang47**
This currently contains no schema validation, only sets output schema to input schema for sources.

### CODE REVIEW

File: cassandra-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/BatchCassandraSource.java [hunk 1]
```diff
@@ -118,6 +119,15 @@ public void transform(KeyValue<Long, Row> input, Emitter<StructuredRecord> emitt
     emitter.emit(builder.build());
   }
 
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(inputSchema);
```
Comment (albertshau) [inline cassandra-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/BatchCassandraSource.java:127]: The input schema for a source will always be null.  Should be setting the schema based on what is set in the config.  Something like:

```
try {
  Schema schema = Schema.parseJson(config.schema);
  pipelineConfigurer.getStageConfigurer().setOutputSchema(schema);
} catch (Exception e) {
  throw new IllegalArgumentException("Invalid schema: " + e.getMessage(), e);
}
```

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/DBSource.java [hunk 2]
```diff
@@ -72,11 +71,6 @@ public DBSource(DBSourceConfig dbSourceConfig) {
   }
 
   @Override
-  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
```
Comment (albertshau) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/DBSource.java:75]: shouldn't remove this

**Comment by albertshau** [comment]
think there was some confusion, sources should be setting outputschema based on their config, not based on the getInputSchema method, which will always return null

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/KVTableSource.java [hunk 3]
```diff
@@ -76,4 +77,11 @@ public KVTableSource(KVTableConfig kvTableConfig) {
   public void transform(KeyValue<byte[], byte[]> input, Emitter<StructuredRecord> emitter) throws Exception {
     emitter.emit(StructuredRecord.builder(SCHEMA).set("key", input.getKey()).set("value", input.getValue()).build());
   }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+
+    // TODO: validate schema
```
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/KVTableSource.java:85]: Why TODO ?

File: core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/SnapshotFileBatchSource.java [hunk 4]
```diff
@@ -51,10 +53,16 @@ public SnapshotFileBatchSource(SnapshotFileSetConfig config) {
...
+    super.configurePipeline(pipelineConfigurer);
+
     PartitionedFileSetProperties.Builder fileProperties = SnapshotFileSet.getBaseProperties(config);
     addFileProperties(fileProperties);
 
     pipelineConfigurer.createDataset(config.getName(), PartitionedFileSet.class, fileProperties.build());
+
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(inputSchema);
+    // TODO: validate schema
```
Comment (nitinmotgi) [inline core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/SnapshotFileBatchSource.java:65]: Remove TODO

**Comment by albertshau** [comment]
closing for now, going to re-open with more complete fixes

**Pull request closed without merging**